src/sqlfluff/rules/L001.py

```diff
@@ -45,7 +45,20 @@ def _eval(self, segment, raw_stack, **kwargs):
             while abs(idx) <= len(raw_stack) and raw_stack[idx].is_type("whitespace"):
                 deletions.append(raw_stack[idx])
                 idx -= 1
-            return LintResult(
-                anchor=deletions[-1], fixes=[LintFix("delete", d) for d in deletions]
+            last_deletion_slice = deletions[-1].pos_marker.source_slice
+
+            # Check the raw source (before template expansion) immediately
+            # following the whitespace we want to delete. Often, what looks
+            # like trailing whitespace in rendered SQL is actually a line like:
+            # "    {% for elem in elements %}\n", in which case the code is
+            # fine -- it's not trailing whitespace from a source code
+            # perspective.
+            next_raw_slice = templated_file.raw_slices_spanning_source_slice(
+                slice(last_deletion_slice.stop, last_deletion_slice.stop)
             )
+            if next_raw_slice[0].slice_type == "literal":
```

Comment:
This appears to work but took a long while for me to understand why!

So are you basically getting all the raw slices which are either templated (`block_start`, `block_end`...etc.) or, if not templated, `literal`? So a "slice" is basically a "templated slice" of either templated language or raw SQL?

If so, it might be nice to explain this:

```suggestion
            # If there are no templated slices, and only raw SQL ("literal"), then safe to delete
            if next_raw_slice[0].slice_type == "literal":
```

Reply:
Close! The way I'm constructing the slice I pass to the function, I know it'll only return one slice, and that is the next slice of raw source (pre-template) immediately following the white space it's considering deleting.

Reply:
I'll add a comment similar to the above. Note there's an existing comment a few lines above which helps explain this as well.

Reply:
Thanks. It's the `literal` that through me for a long time. I was wondering why you weren't checking for `"\n"` rather than any literal. But get it now - `literal` here means non-templated code, rather than any particular literal value.

Reply:
Yes, it's literal in the template sense, not in the SQL sense. If you think about what Jinja or other templating engines do, they execute the control structures, and other stuff just passes through. If you mentally translate the template into the code you'd write by hand if you _didn't_ have Jinja, "other stuff" would be a `print()` statement, e.g. the string "0" in a template becomes:
```
print('0')
```

when rendering the template.